CryptoOnramp SDK: Adds KYC Refresh API + UI

Twigz opened the pull request:
# Summary
Adds the new KYC Refresh API to CryptoOnrampCoordinator, with supporting UI in StripePaymentSheet to reduce the amount of public exposures needed in StripePaymentSheet.

There are several remaining items not included here, which will be added in follow-up PRs:

Necessary Localizations
Adding analytics to the crypto onramp implementation of the new API.
Handling the new API’s result properly in the demo app, specifically presenting a new UI to re-collect the address to reinvoke the API to confirm the changes.

# Motivation
[Product requirements](https://docs.google.com/document/d/1ZLfTbJljlyaJS_G16X98aZQqeghp6f3flxhNqE7uqHI/edit?tab=t.c0547slx70xm#heading=h.xg0apl9hsn12)
[Mobile API Review draft](https://docs.google.com/document/d/15pjTJiAn5pcajm0Z820A4yLt517q0l5wOVxu7YnLkaM/edit?tab=t.0#heading=h.570u4leviueu)


# Testing
<!-- How was the code tested? Be as specific as possible. -->
- [ ] Added tests
- [ ] Modified tests
- [x] Manually verified

1. Logged in with a user in the demo app.
2. After authenticating, used the new Verify KYC Info button..
Separately chose the 3 options to dismiss the UI (x button, pencil icon, and confirm button) ensuring that the proper status was reached in the demo UI code / print statements, and that the API requests succeeded / failed where appropriate.

<!-- Ignored Tests Did you newly ignore a test in this PR?  If so, please open an R4 incident so that the test can be re-enabled as soon as possible-->

# Screenshots

| Light | Dark |
| --- | --- |
| <img src="https://github.com/user-attachments/assets/60cf606f-0e4c-4ed1-8c65-d34fffcef704" width="300" /> | <img src="https://github.com/user-attachments/assets/1682032a-4ced-45fc-9555-36ce02c48a50" width="300" /> |

# Changelog
<!-- Is this a notable change that affects users? If so, add a line to `CHANGELOG.md` and prefix the line with one of the following:
    - [Added] for new features.
    - [Changed] for changes in existing functionality.
    - [Deprecated] for soon-to-be removed features.
    - [Removed] for now removed features.
    - [Fixed] for any bug fixes.
    - [Security] in case of vulnerabilities.
-->

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/OnrampPresenterCoordinator.kt:
```diff
@@ -110,6 +175,28 @@ internal class OnrampPresenterCoordinator @Inject constructor(
+        coroutineScope.launch {
+            consumerSessionClientSecret()?.let {
+                val result = cryptoApiRepository.retrieveKycInfo(it)
+                if (result.isSuccess) {
```
use `fold` for better compile time safety:
```kotlin
                result.fold(
                    onSuccess = TODO(),
                    onFailure = TOOO()
                )
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/OnrampPresenterCoordinator.kt:
```diff
@@ -56,6 +67,52 @@ internal class OnrampPresenterCoordinator @Inject constructor(
     private val currentLinkAccount: LinkController.LinkAccount?
         get() = interactor.state.value.linkControllerState?.internalLinkAccount
 
+    private val onrampActivityResultLauncher: ActivityResultLauncher<VerifyKycActivityContractArgs> =
```
```suggestion
    private val verifyKycResultLauncher: ActivityResultLauncher<VerifyKycActivityContractArgs> =
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/OnrampPresenterCoordinator.kt:
```diff
@@ -32,6 +42,7 @@ internal class OnrampPresenterCoordinator @Inject constructor(
     linkController: LinkController,
     lifecycleOwner: LifecycleOwner,
     private val activity: ComponentActivity,
+    private val cryptoApiRepository: CryptoApiRepository,
```
I think adding this dependency as well as pulling the CSCS from the interactor are code smells. We should aim to keep this Coordinator class dumb (like an Activity/View -- mostly wiring) and move the logic to `OnrampInteractor`.

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/OnrampPresenterCoordinator.kt:
```diff
@@ -56,6 +67,52 @@ internal class OnrampPresenterCoordinator @Inject constructor(
 
+    private val onrampActivityResultLauncher: ActivityResultLauncher<VerifyKycActivityContractArgs> =
+        activity.registerForActivityResult(VerifyKycInfoActivityContract()) { result ->
+            when (val action = result.action) {
```
re: moving logic to the interactor

Essentially this entire `when` block should be moved to `OnrampInteractor`, say method 
```kotlin
suspend fun onVerifyKycResult(
    result: VerifyKycActivityContractResult
): OnrampVerifyKycInfoResult
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/OnrampPresenterCoordinator.kt:
```diff
@@ -202,6 +293,9 @@ internal class OnrampPresenterCoordinator @Inject constructor(
         }
     }
 
+    private fun appearance(): LinkAppearance? =
```
same here re: interactor

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/model/DateOfBirthSerializer.kt:
```diff
@@ -0,0 +1,35 @@
+    override fun deserialize(decoder: Decoder): DateOfBirth {
+        val surrogate = delegate.deserialize(decoder)
+
+        return DateOfBirth(surrogate.day, surrogate.month, surrogate.year)
```
add names to call args (especially important here)
```suggestion
        return DateOfBirth(surrogate.day, surrogate.month, surrogate.year)
```

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/KYCRefreshScreen.kt:
```diff
@@ -0,0 +1,228 @@
+@Composable
+@Suppress("LongMethod")
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+fun KYCRefreshScreen(
```
code style:
```suggestion
fun KycRefreshScreen(
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/model/KycRetrieveResponse.kt:
```diff
@@ -10,20 +14,22 @@ import kotlinx.serialization.Serializable
-    val address: PaymentSheet.Address
-)
+    override val address: PaymentSheet.Address
+) : Parcelable, VerifyKYCInfo
```
`VerifyKYCInfo` is a view-layer interface that shouldn't be applied to data models. We should be using converter methods instead, e.g. 
```kotlin
fun KycRetrieveResponse.toVerifyKYCInfo(): VerifyKYCInfo
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/ui/VerifyKycInfoActivity.kt:
```diff
@@ -0,0 +1,88 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val kycInfo = intent.getParcelableExtra<KycRetrieveResponse>("kycInfo")!!
+        val linkAppearance = intent.getParcelableExtra<LinkAppearance>("linkAppearance")
```
1. avoid using these deprecated methods (see what we do elsewhere)
2. extract these key values to consts especially since they need to be in sync with `createIntent()`

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/ui/VerifyKycInfoActivity.kt:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    private fun createIntent(action: KycRefreshScreenAction): Intent {
```
rename
```suggestion
    private fun createResultIntent(action: KycRefreshScreenAction): Intent {
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/ui/VerifyKycInfoActivity.kt:
```diff
@@ -0,0 +1,88 @@
+)
+
+@Parcelize
+internal sealed class KycRefreshScreenAction : Parcelable {
```
```suggestion
internal sealed interface KycRefreshScreenAction : Parcelable {
```

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/KYCRefreshScreen.kt:
```diff
@@ -0,0 +1,228 @@
+                            InfoRow(title = "Name", value = name)
+                            Divider(color = LinkTheme.colors.textPrimary.copy(alpha = 0.12f))
+                            InfoRow(title = "Date of Birth", value = dob)
+                            Divider(color = LinkTheme.colors.textPrimary.copy(alpha = 0.12f))
```
Extract out divider component

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/KYCRefreshScreen.kt:
```diff
@@ -0,0 +1,228 @@
+                    .fillMaxSize()
+                    .background(LinkTheme.colors.surfacePrimary)
+            ) {
+                Column(modifier = Modifier.padding(24.dp)) {
```
This needs to be vertically scrollable

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/KYCRefreshScreen.kt:
```diff
@@ -0,0 +1,228 @@
+}
+
+private fun PaymentSheet.Address.formattedAddress(): String {
+    return listOf(
```
How do we format addresses elsewhere? I feel like this should already exist.

Twigz replied:
I thought so as well, but the closest I could find was in `AddressElementExampleActivity`, and it's not quite the same. Happy to be proven wrong though.

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/KYCRefreshScreen.kt:
```diff
@@ -0,0 +1,228 @@
+                    .background(LinkTheme.colors.surfacePrimary)
+            ) {
+                Column(modifier = Modifier.padding(24.dp)) {
+                    TopNavigationBar(
```
Can't we use `LinkAppBar`?

lng-stripe commented on paymentsheet/res/drawable/stripe_ic_kyc_verify_edit_day.xml:
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Can't we use `stripe_ic_edit_outlined_symbol` and tint it appropriately?

Twigz commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/KYCRefreshScreen.kt:
```diff
@@ -0,0 +1,228 @@
+    onConfirm: () -> Unit
+) {
+    val name = "${kycInfo.firstName} ${kycInfo.lastName}"
+    val dob = "%02d/%02d/%d".format(kycInfo.dateOfBirth.month, kycInfo.dateOfBirth.day, kycInfo.dateOfBirth.year)
```
I did push up a localization of this, but I wasn't able to find a good solution for the following to match the design:
1. Getting preceding 0's for the day/month if the numbers were single digits
2. Displaying 4 digits for the year

With the restriction of the minSdk being 21, I didn't see any available APIs that would let me keep the Locale considered while accounting for those two.

Are you aware of any / should I fallback to what I've now implemented for SDK 21 and use more modern APIs when they're available?

lng-stripe commented on paymentsheet/res/values/strings.xml:
```diff
@@ -72,6 +72,18 @@ to be saved and used in future checkout sessions. -->
+  <!-- Title for the name field on the KYC verification screen -->
+  <string name="stripe_link_onramp_kyc_verification_name_field_title">Name</string>
+  <!-- Title for the date of birth field on the KYC verification screen -->
+  <string name="stripe_link_onramp_kyc_verification_dob_field_title">Date of Birth</string>
```
I'm pretty sure this should be sentence case per our style guide.
```suggestion
  <string name="stripe_link_onramp_kyc_verification_dob_field_title">Date of birth</string>
```

lng-stripe commented on paymentsheet/res/values/strings.xml:
```diff
@@ -72,6 +72,18 @@ to be saved and used in future checkout sessions. -->
   <!-- Menu option to update card on the wallet screen on Link native -->
   <string name="stripe_link_wallet_menu_action_update_card">Update card</string>
+  <!-- Title for the name field on the KYC verification screen -->
+  <string name="stripe_link_onramp_kyc_verification_screen_title">Confirm your information</string>
```
Description doesn't match

lng-stripe commented:
A few issues in the demo below:
 1. Confirm button is covered by the nav bar. Need to add bottom safe content padding.
 2. The dismissal animation should slide down with the bottom sheet animation, not fade out like an Activity. I think we fix this either by:
    a. making dismissal actions dismiss the bottom sheet and add a listener to finish the activity when the bottom sheet has dismissed, or
    b. dismiss the bottom sheet and add a artificial delay before Activity `finish()`
 3. The edit icon still seems too inset left. Left and right margins should be identical, but let me know if design disagrees.

https://github.com/user-attachments/assets/f1708410-6c56-4f1b-84c1-a4096279aa03

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/model/OnrampVerifyKycInfoResult.kt:
```diff
@@ -0,0 +1,38 @@
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+sealed interface OnrampVerifyKycInfoResult {
+    /**
+     * Kyc verified successfully.
```
nit: "KYC" here and below in docstrings 
```suggestion
     * KYC verified successfully.
```

lng-stripe commented on crypto-onramp/src/main/java/com/stripe/android/crypto/onramp/ui/VerifyKycInfoActivity.kt:
```diff
@@ -0,0 +1,136 @@
+        enableEdgeToEdge()
+
+        setContent {
+            OnrampKYCRefreshScreen(
```
nit:
```suggestion
            OnrampKycRefreshScreen(
```

Twigz commented:
> A few issues in the demo below:
> 
> 1. Confirm button is covered by the nav bar. Need to add bottom safe content padding.

Solved

> 2. The dismissal animation should slide down with the bottom sheet animation, not fade out like an Activity. I think we fix this either by:
>    a. making dismissal actions dismiss the bottom sheet and add a listener to finish the activity when the bottom sheet has dismissed, or
>    b. dismiss the bottom sheet and add a artificial delay before Activity `finish()`

This was actually happening in most cases, just not the hardware back button, so I added a BackHandler to account for that.

> 3. The edit icon still seems too inset left. Left and right margins should be identical, but let me know if design disagrees.

I adjusted this over to account for the extra padding being automatically added by the system on top of our 16.dp inset. As is I think it makes sense because it keep the tap target where the user expects it to be (container within the box).

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/onramp/ui/OnrampKYCRefreshScreen.kt:
```diff
@@ -0,0 +1,268 @@
+            .fillMaxWidth()
+            .padding(vertical = 16.dp)
+            .padding(start = 16.dp)
+            .padding(end = if (icon == null) 16.dp else 0.dp),
```
nit: combine into 1 `padding()` call
